net/rhp/program.go

```diff
@@ -0,0 +1,633 @@
+	// memory, here I've opted to use only the instruction's memory.
+	costs.BaseCost = writeCost(settings, SectorSize).Add(resourceCost(settings, costs.Memory, costs.Time))
+	// storage cost is the cost of storing 1 sector for the remaining duration.
+	costs.StorageCost = settings.StoragePrice.Mul64(SectorSize * duration)
+	// additional collateral is the collateral the host is expected to put up
+	// per sector per block.
+	// note: in siad the additional collateral does not consider remaining
+	// duration.
+	costs.AdditionalCollateral = settings.Collateral.Mul64(SectorSize * duration)
+	return
+}
+
+// DropSectorsCost returns the cost of the drop sectors instruction.
+func DropSectorsCost(settings HostSettings, n uint64) (costs ResourceUsage) {
+	costs.BaseCost = settings.InstrDropSectorsUnitCost.Mul64(n).Add(settings.InstrDropSectorsBaseCost)
+	return
+}
+
+// HasSectorCost returns the cost of the has sector instruction.
+func HasSectorCost(settings HostSettings) (costs ResourceUsage) {
+	costs.BaseCost = settings.InstrHasSectorBaseCost
+	return
+}
+
+// ReadCost returns the cost of the read instruction.
+func ReadCost(settings HostSettings, l uint64) (costs ResourceUsage) {
+	costs.BaseCost = settings.ProgReadCost.Mul64(l).Add(settings.InstrReadBaseCost)
+	return
+}
+
+// RevisionCost returns the cost of the revision instruction.
+func RevisionCost(settings HostSettings) (costs ResourceUsage) {
+	costs.BaseCost = settings.InstrRevisionBaseCost
+	return
+}
+
+// SwapSectorCost returns the cost of the swap sector instruction.
+func SwapSectorCost(settings HostSettings) (costs ResourceUsage) {
+	costs.BaseCost = settings.InstrSwapSectorBaseCost
+	return
+}
+
+// UpdateRegistryCost returns the cost of the update registry instruction.
+func UpdateRegistryCost(settings HostSettings) (costs ResourceUsage) {
+	costs.BaseCost = writeCost(settings, 256).Add(settings.InstrUpdateRegistryBaseCost)
+	// storing 256 bytes for 5 years
+	costs.StorageCost = settings.StoragePrice.Mul64(256 * 5 * blocksPerYear)
+	return
+}
+
+// ReadRegistryCost returns the cost of the read registry instruction.
+func ReadRegistryCost(settings HostSettings) (costs ResourceUsage) {
+	costs.BaseCost = writeCost(settings, 256).Add(settings.InstrReadRegistryBaseCost)
+	// storing 256 bytes for 10 years
+	costs.StorageCost = settings.StoragePrice.Mul64(256 * 10 * blocksPerYear)
+	return
+}
+
+func writeInstruction(e *types.Encoder, instr Instruction) {
+	specifier := instr.Specifier()
+	e.Write(specifier[:])
+	switch v := instr.(type) {
+	case InstrAppendSector:
+		v.EncodeTo(e)
+	case InstrDropSectors:
+		v.EncodeTo(e)
+	case InstrHasSector:
+		v.EncodeTo(e)
+	case InstrReadOffset:
+		v.EncodeTo(e)
+	case InstrReadSector:
+		v.EncodeTo(e)
+	case InstrContractRevision:
+		v.EncodeTo(e)
+	case InstrSwapSector:
+		v.EncodeTo(e)
+	case InstrUpdateRegistry:
+		v.EncodeTo(e)
+	case InstrReadRegistry:
+		v.EncodeTo(e)
```

Comment:
we could either add `rpc.Object` to the `Instruction` interface, or you could use a type assertion:
```go
et, ok := instr.(types.EncoderTo)
if !ok {
    panic(fmt.Sprintf("%T does not implement types.EncoderTo", v))
}
et.EncodeTo(e)
```